### DISCUSSION TITLE
Is forcing Google to a specific tld for the language still matter?

### DISCUSSION THREAD

**Discussion by unixfox** [searx/searx#2506] [Development]
I just tried to force every request to www.google.com, and I still get the result in the correct language:
```python
    query_url = 'https://www.google.com/search' + "?" + urlencode({
        'q': query,
        'hl': lang_country,
        'lr': "lang_" + language,
        'ie': "utf8",
        'oe': "utf8",
        'start': offset,
    })
```
Related to https://github.com/searx/searx/blob/master/searx/engines/google.py#L175

Is it still important to use the correct tld (.de or .fr) for having the results in the correct language?

@return42 @dalf What are your thoughts on this? Could we remove this part of the code (https://github.com/searx/searx/blob/master/searx/engines/google.py#L41-91) and just use www.google.com?

**Comment by return42** [comment]
> Is it still important to use the correct tld (.de or .fr) for having the results in the correct language?

I don't know ... I implemented TLDs because it was obvious to me ... I haven't compared the result qualities across TLDs.

Thanks for comparing TLDs with .com  .. brings me to a complete new ideas ...

Today I implemented https://github.com/searx/searx/pull/2482/commits/890839df4 and made a lot of tests automated by the search.checker matrix from @dalf.  While I was running the tests I realized that I got captcha very fast on google.com.

I guess it was because the tests of google-images, google-videos and google (for en-US) are all going to google.com.

Back to your question: Even if the result quality is the same on google.com for the different languages, using TLDs can help to reduce the requests on google.com and this reduce the risk to get a CAPTCHA.

And now a competed new idea:

If the result qualities are all the same on the different TLDs we can rotate the TDLs to reduce google CAPTCHA.

MIgth be worth to make some analysis .. it would be important to know whether the quality of the results are not degrading. 

.. what do you think @dalf @unixfox

**Reply by unixfox** [reply]
> > Is it still important to use the correct tld (.de or .fr) for having the results in the correct language?
> 
> I don't know ... I implemented TLDs because it was obvious to me ... I haven't compared the result qualities across TLDs.

I don't think it really matters for the results because for example www.google.de and www.google.be are using the same IP address. I think the tld (.be or .de) is just a gimmick and doesn't really serve a different kind of "Google code".

> Thanks for comparing TLDs with .com .. brings me to a complete new ideas ...
> 
> Today I implemented [890839d](https://github.com/searx/searx/commit/890839df4) and made a lot of tests automated by the search.checker matrix from @dalf. While I was running the tests I realized that I got captcha very fast on google.com.
> 
> I guess it was because the tests of google-images, google-videos and google (for en-US) are all going to google.com.
> 
> Back to your question: Even if the result quality is the same on google.com for the different languages, using TLDs can help to reduce the requests on google.com and this reduce the risk to get a CAPTCHA.

Are you sure about that? www.google.be and www.google.de are for example using the same IP address (at least in IPv6). I don't think it really helps for the CAPTCHA because when your IP is blocked, switching to other languages just won't help to unblock it.

> And now a competed new idea:
> 
> If the result qualities are all the same on the different TLDs we can rotate the TDLs to reduce google CAPTCHA.
> 
> MIgth be worth to make some analysis .. it would be important to know whether the quality of the results are not degrading.
> 
> .. what do you think @dalf @unixfox

See my previous comment about the CAPTCHA, I don't think it will make a difference. You should prove your facts because I don't see how we can escape recaptcha by just rotating the tld.

**Comment by return42** [comment]
> Are you sure about that? www.google.be and www.google.de are for example using the same IP address (at least in IPv6). I

I did not't tested IPs nor captcha, in my [test](https://github.com/searx/searx/pull/2482/files#diff-e95e3f454925bc2344d3cf538950b11a1255935b47a609c9589e6ac33d74802fR495-R496) I tested language results for  [fe, de, en, and zn-CH](https://github.com/searx/searx/pull/2482/files#diff-e95e3f454925bc2344d3cf538950b11a1255935b47a609c9589e6ac33d74802fR132-R140) and only the en gets blocked .. so there is a difference .. 

>  You should prove your facts because I don't see how we can escape recaptcha by just rotating the tld.

It is a discussion and I only said that it might be worth investigating more closely: I see an effect where google.com blocks me while other TLDs response results, at the same time to the same query ... and this is a kind of escaping from CAPTCHA .. at least for me.

Rotating TLDs could be a method to minimize the CAPTCHA probability .. even if some of TLDs are pointing to the same IP, i think it is worth to investigate closer.
